### PR TITLE
Add reason to hash

### PR DESCRIPTION
**Pull request by camillavk**

### CODE REVIEW

File: lib/postcode_validation/use_case/validate_address.rb [hunk 1]
```diff
@@ -11,10 +11,18 @@ def initialize(address_match_gateway:, logger: nil)
       def execute(postcode:, country:)
         @postcode = postcode
         @country = country
-        { valid?: valid_postcode? }
+        errors = []
+        errors << 'no_country_provided' if country.nil?
+        errors << 'invalid_format' if invalid_postcode_format?
+        matches = potential_address_matches
+        errors << 'no_matches' if matches.first.nil?
+        return { valid?: false, reason: errors } unless errors.empty?
```
Comment (emileswarts) [inline lib/postcode_validation/use_case/validate_address.rb]: Wonder if we could push some of this stuff down into methods.  Can help you check it out if you want.

File: lib/postcode_validation/use_case/validate_address.rb [hunk 2]
```diff
@@ -6,42 +6,59 @@ def initialize(address_match_gateway:, logger: nil)
...
 
-      def potential_address_matches
-        @address_match_gateway.query(search_term: postcode,
-                                     country: country)
+      def ensure_country(country)
+        @errors << 'no_country_provided' if country.nil?
       end
 
-      def invalid_postcode_format?
+      def ensure_postcode_format(postcode, country)
```
Comment (craigjbass) [inline lib/postcode_validation/use_case/validate_address.rb]: `check_postcode_format`

File: lib/postcode_validation/use_case/validate_address.rb [hunk 3]
```diff
@@ -6,42 +6,59 @@ def initialize(address_match_gateway:, logger: nil)
...
-        matches.each { |address| return true if address.postcode_matches? postcode }
-        false
+      def ensure_results(result)
+        @errors << 'no_matches' if result.first.nil?
       end
 
-      def potential_address_matches
-        @address_match_gateway.query(search_term: postcode,
-                                     country: country)
+      def ensure_country(country)
```
Comment (craigjbass) [inline lib/postcode_validation/use_case/validate_address.rb]: `check_country`

File: lib/postcode_validation/use_case/validate_address.rb [hunk 4]
```diff
@@ -6,42 +6,59 @@ def initialize(address_match_gateway:, logger: nil)
...
-        matches = potential_address_matches
-        return false if matches.first.nil?
+      def matched_addresses(postcode, country)
+        @address_match_gateway.query(search_term: postcode,
+                                     country: country)
+      end
 
-        matches.each { |address| return true if address.postcode_matches? postcode }
-        false
+      def ensure_results(result)
```
Comment (craigjbass) [inline lib/postcode_validation/use_case/validate_address.rb]: `check_matched_addresses`

File: lib/postcode_validation/use_case/validate_address.rb [hunk 5]
```diff
@@ -6,42 +6,59 @@ def initialize(address_match_gateway:, logger: nil)
...
 
       private
 
       attr_reader :address_match_gateway, :logger, :postcode, :country
 
+      def result_payload(result, postcode)
+        return { valid?: false, reason: @errors } unless @errors.empty?
+
+        result.each do |address|
+          return { valid?: true, reason: 'valid_postcode' } if address.postcode_matches? postcode
```
Comment (craigjbass) [inline lib/postcode_validation/use_case/validate_address.rb]: reason: should always be an array. string is liskov incompatible with array which makes this all hell harder to consume in static-typed languages

**Comment by camillavk** [comment]
@craigjbass @emileswarts updated

**Pull request merged**